time start_time
time end_time
point_2d[] p
